cedar-policy-validator/src/lib.rs

```diff
@@ -206,6 +286,194 @@ impl Validator {
+        match v.first() {
+            Some(e) => {
+                let mut p = e.clone();
+                for e in v {
+                    if e.0 < p.0 {
+                        p = e.clone();
+                    }
+                }
+                p
+            }
+            None => (EntityDerefLevel { level: Some(0) }, None),
+        }
+    }
+
+    #[cfg(feature = "level-validate")]
+    ///Walk the type-annotated AST and compute the used level and possible violation
+    fn check_entity_deref_level_helper<'a>(
+        &'a self,
+        e: &cedar_policy_core::ast::Expr<Option<crate::types::Type>>,
+        max_allowed_level: &EntityDerefLevel,
+        policy_id: &PolicyID,
+    ) -> (EntityDerefLevel, Option<EntityDerefLevelViolation>) {
+        use crate::types::{EntityRecordKind, Type};
+        use cedar_policy_core::ast::ExprKind;
+        match e.expr_kind() {
+            ExprKind::Lit(_) => (
+                EntityDerefLevel { level: Some(0) }, //Literals can't be dereferenced
+                None,
+            ),
+            ExprKind::Var(_) => (max_allowed_level.clone(), None), //Roots start at `max_allowed_level`
+            ExprKind::Slot(_) => (EntityDerefLevel { level: Some(0) }, None), //Slot will be replaced by Entity literal so treat the same
+            ExprKind::Unknown(_) => (
+                EntityDerefLevel { level: Some(0) }, //Can't dereference an unknown
+                None,
+            ),
+            ExprKind::If {
+                test_expr,
+                then_expr,
+                else_expr,
+            } => {
+                let es = [test_expr, then_expr, else_expr];
+                let v: Vec<(EntityDerefLevel, Option<_>)> = es
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::And { left, right } | ExprKind::Or { left, right } => {
+                let es = [left, right];
+                let v: Vec<(EntityDerefLevel, Option<_>)> = es
+                    .iter()
+                    .map(|l| self.check_entity_deref_level_helper(l, max_allowed_level, policy_id))
+                    .collect();
+                Self::min(v)
+            }
+            ExprKind::UnaryApp { arg, .. } => {
+                self.check_entity_deref_level_helper(arg, max_allowed_level, policy_id)
+            }
+            // `In` operator decrements the LHS only
+            ExprKind::BinaryApp { op, arg1, arg2 } if op == &BinaryOp::In => {
+                let mut lhs =
+                    self.check_entity_deref_level_helper(arg1, max_allowed_level, policy_id);
+                let rhs = self.check_entity_deref_level_helper(arg2, max_allowed_level, policy_id);
+                lhs = (lhs.0.decrement(), lhs.1);
+                let new_level = Self::min(vec![lhs, rhs]).0;
+                if new_level.level < Some(0) {
```

Comment:
Is `rhs` relevant to this comparison?

Reply:
right, I'm wondering if it could ever cause `min` to be less than 0 when there wasn't already an error raised earlier

Reply:
I'm not sure I understand. Shouldn't the level of `principal.foo.bar.baz + resource.foo.bar.baz.biz` should consider the rhs? The most severe error is "propagated" up the expression tree using `min` so we can report the actual level.